sdk/python/feast/staging/storage_client.py

```diff
@@ -0,0 +1,312 @@
+            remote_path (str): relative path to the folder to which the files need to be uploaded
+        """
+        protocol = self._get_staging_protocol(scheme)
+        return protocol.upload_file(local_path, bucket, remote_path)
+
+    def _get_staging_protocol(self, protocol):
+        """
+        Lazy initialization of a specific protocol object(GCSProtocol, S3Protocol etc.) w
+        hen the 1st instance is encountered.
+
+        Args:
+            protocol (str): uri scheme: s3, gs or file
+
+        Returns:
+            An object of concrete implementation of AbstractStagingProtocol
+        """
+        if protocol in self._protocol_dict:
+            return self._protocol_dict[protocol]
+        else:
+            if PROTOCOL(protocol) == PROTOCOL.GS:
+                self._protocol_dict[protocol] = GCSProtocol()
+            elif PROTOCOL(protocol) == PROTOCOL.S3:
+                self._protocol_dict[protocol] = S3Protocol()
+            elif PROTOCOL(protocol) == PROTOCOL.LOCAL_FILE:
+                self._protocol_dict[protocol] = LocalFSProtocol()
+            else:
+                raise Exception(
+                    f"Could not identify file protocol {protocol}. Only gs:// and file:// and s3:// supported"
+                )
+            return self._protocol_dict[protocol]
+
+
+class AbstractStagingProtocol(ABC):
+    """
+    Abstract class for staging protocol, any new protocol should implement this class.
+    """
+
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def download_file(self, uri: ParseResult) -> TemporaryFile:
+        pass
+
+    @abstractmethod
+    def list_files(self, bucket: str, uri: ParseResult) -> List[str]:
+        pass
+
+    @abstractmethod
+    def upload_file(self, local_path: str, bucket: str, remote_path: str):
+        pass
+
+
+class GCSProtocol(AbstractStagingProtocol):
+    """
+    Implementation of AbstractStagingProtocol for google cloud storage
+    """
+
+    def __init__(self):
+        try:
+            from google.cloud import storage
+        except ImportError:
+            raise ImportError(
+                "Install package google-cloud-storage==1.20.* for gcs staging support"
+                "run ```pip install google-cloud-storage==1.20.*```"
+            )
+        self.gcs_client = storage.Client(project=None)
+
+    def download_file(self, uri: ParseResult) -> TemporaryFile:
+        """
+        Downloads a file from google cloud storage and returns a TemporaryFile object
+
+        Args:
+            uri (urllib.parse.ParseResult): Parsed uri of the file ex: urlparse("gs://bucket/file.avro")
+
+        Returns:
+             TemporaryFile object
+        """
+        url = uri.geturl()
+        file_obj = TemporaryFile()
+        self.gcs_client.download_blob_to_file(url, file_obj)
+        return file_obj
+
+    def list_files(self, bucket: str, uri: ParseResult) -> List[str]:
```

Comment:
Is there an example of `list_files` being tested or used? Are we enforcing the use of wildcards? Normally people dont use wildcards for listing objects on GCS.

Reply:
This was an existing piece of code which has been moved here. (`def _get_files(bucket: str, uri: ParseResult) -> List[str]:` method in `file.py`).